File: src/libs/plane.cxx
#include <tiger/toolkit/plane.h>
#include <tiger/utils.h>
#include <iostream>

toolkit::Plane::Plane() : m_N(0.0, 1.0, 0.0), m_P(0.0, 0.0, 0.0) 
{
}

toolkit::Plane::Plane(
    const Eigen::Vector3d & N, 
    const Eigen::Vector3d & P, 
    bool normalize, 
    bool fixZeros, 
    double threshold) : m_N(N), m_P(P)
{
    if (normalize)
    {
        m_N.normalize();
    }

    if (fixZeros) 
    {
        utils::fixZeros(m_N, threshold);
        utils::fixZeros(m_P, threshold);
    }
}

toolkit::Plane::Plane(
    double nx, 
    double ny, 
    double nz, 
    double px, 
    double py, 
    double pz, 
    bool normalize, 
    bool fixZeros, 
    double threshold) : m_N(nx, ny, nz), m_P(px, py, pz)
{
    if (normalize)
    {
        m_N.normalize();
    }

    if (fixZeros) 
    {
        utils::fixZeros(m_N, threshold);
        utils::fixZeros(m_P, threshold);
    }
}

toolkit::Plane::~Plane() 
{
}

toolkit::Plane toolkit::Plane::Clone() const
{
    return Plane(m_N, m_P);
}

bool toolkit::Plane::IsPointInPlane(const Eigen::Vector3d & Q, double threshold) const 
{
    return PointLocation(Q, threshold) == 0;
}

const Eigen::Vector3d & toolkit::Plane::Normal() const 
{
    return m_N;
}

void toolkit::Plane::Normal(const Eigen::Vector3d & N, bool normalize)
{
    m_N << N;

    if (normalize)
    {
        m_N.normalize();
    }
}

void toolkit::Plane::Normal(double x, double y, double z, bool normalize)
{
    m_N << x, y, z;

    if (normalize)
    {
        m_N.normalize();
    }
}

const Eigen::Vector3d & toolkit::Plane::Point() const 
{
    return m_P;
}

void toolkit::Plane::Point(const Eigen::Vector3d & P)
{
    m_P << P;
}

void toolkit::Plane::Point(double x, double y, double z)
{
    m_P << x, y, z;
}

int toolkit::Plane::PointLocation(const Eigen::Vector3d & Q, double threshold) const
{
    // Calculate the dot product between the normal vector and the PQ vector. Then, return its 
    // value (check if it is close to zero)
    double dot = m_N.dot(Q - m_P);
    return (abs(dot) <= threshold) ? 0 : ((dot > 0.0) ? 1 : -1);
}

void toolkit::Plane::Set(const Eigen::Vector3d & N, const Eigen::Vector3d & P, bool normalize)
{
    m_N << N;
    m_P << P;

    if (normalize)
    {
        m_N.normalize();
    }
}

void toolkit::Plane::Set(
    double nx, 
    double ny, 
    double nz, 
    double px, 
    double py, 
    double pz, 
    bool normalize) 
{
    m_N << nx, ny, nz;
    m_P << px, py, pz;

    if (normalize)
    {
        m_N.normalize();
    }
}

void toolkit::Plane::Write() const 
{
    std::cout << "N = (" << m_N.x() << ", " << m_N.y() << ", " << m_N.z() << ")" << std::endl;
    std::cout << "P = (" << m_P.x() << ", " << m_P.y() << ", " << m_P.z() << ")" << std::endl;
}

double toolkit::Plane::a() const
{
    return m_N.x();
}

double toolkit::Plane::b() const
{
    return m_N.y();
}

double toolkit::Plane::c() const
{
    return m_N.z();
}

double toolkit::Plane::d() const
{
    return m_N.dot(m_P);
}

File: python/toolkit/plane.py
# -*- coding: utf-8 -*-

from __future__ import annotations
from toolkit.pvector import PVector

# The class to represent a plane in 3D. A plane is defined by a normal vector N
# and a point P, the implicit definition of the plane is N * (Q - P) = 0, where
# Q is any point in the plane and * is the dot product.
class Plane:
    
    # Constructor of the class.
    def __init__(self, N:PVector=PVector(), P:PVector=PVector()):
        assert N is not None, "Missing N"
        assert isinstance(N, PVector), "N not a PVector"
        assert P is not None, "Missing P"
        assert isinstance(P, PVector), "P not a PVector"
        self.normal = N.copy()
        self.point = P.copy()
    
    
    def __getattr__(self, name:str):
        return self.__dict__[f"{name}"]
    
    
    def __repr__(self):
        return f"Plane:(normal:{self.normal}, point:{self.point})"
    
    
    def __setattr__(self, name:str, value):
        assert isinstance(value, PVector), "value not a PVector"
        self.__dict__[f"{name}"] = value
    
    
    def __str__(self):
        return f"(normal:{self.normal}, point:{self.point})"
    
    
    def clone(self) -> Plane:
        return self.copy()
    
    
    def copy(self) -> Plane:
        return Plane(self.normal, self.point)
    
    
    def fixzeros(self, e:int|float=1e-8) -> Plane:
        assert isinstance(e, (int, float)), "e not a real number"
        self.normal.fixzeros(e)
        self.point.fixzeros(e)
        return self
    
    
    def ispoint(self, Q:PVector, e:int|float=1e-8, r:int=10) -> bool:
        assert Q is not None, "Missing Q"
        assert isinstance(Q, PVector), "Q not a PVector"
        assert isinstance(e, (int, float)), "e not a real number"
        assert isinstance(r, int), "r not an integer number"
        return self.pointlocation(Q, e, r) == 0
    
    
    # Checks the location of a point Q with respect of the plane. A 1 value 
    # is in the positive half space of the plane (in the direction of the 
    # normal vector), a 0 value indicates the point is at the plane, a -1 value
    # indicates the point is at the negative half space of the plane (in the 
    # opposite direction of the normal vector). The indicator comes from 
    # Dot(N, Q - P).
    def pointlocation(self, Q:PVector, e:int|float=1e-8, r:int=10) -> int:
        assert Q is not None, "Missing Q"
        assert isinstance(Q, PVector), "Q not a PVector"
        assert isinstance(e, (int, float)), "e not a real number"
        assert isinstance(r, int), "r not an integer number"
        dot = round(self.normal.dot(Q - self.point), r)
        return 0 if abs(dot) <= e else (1 if dot > 0 else -1)
    
    
    # Returns the values a, b, c, d representing the coefficients of the 
    # equation ax + by + cz = d.
    def equationcoefficients(self) -> tuple:
        return self.normal.x, self.normal.y, self.normal.z, self.normal.dot(self.point)
    
    
    def set(self, nx:int|float, ny:int|float, nz:int|float, px:int|float, py:int|float, pz:int|float) -> Plane:
        assert isinstance(nx, (int, float)), "nx not a real number"
        assert isinstance(ny, (int, float)), "ny not a real number"
        assert isinstance(nz, (int, float)), "nz not a real number"
        assert isinstance(px, (int, float)), "px not a real number"
        assert isinstance(py, (int, float)), "py not a real number"
        assert isinstance(pz, (int, float)), "pz not a real number"
        self.normal.set(nx, ny, nz)
        self.point.set(px, py, pz)
        return self
    
    
    def setnormal(self, x:int|float, y:int|float, z:int|float) -> Plane:
        assert isinstance(x, (int, float)), "x not a real number"
        assert isinstance(y, (int, float)), "y not a real number"
        assert isinstance(z, (int, float)), "z not a real number"
        self.normal.set(x, y, z)
        return self
    
    
    def setpoint(self, x:int|float, y:int|float, z:int|float) -> Plane:
        assert isinstance(x, (int, float)), "x not a real number"
        assert isinstance(y, (int, float)), "y not a real number"
        assert isinstance(z, (int, float)), "z not a real number"
        self.point.set(x, y, z)
        return self
    